fl4health/clients/model_merge_client.py

```diff
@@ -0,0 +1,292 @@
+        """
+        self.model = self.get_model(config)
+        self.test_loader = self.get_test_data_loader(config)
+        self.parameter_exchanger = self.get_parameter_exchanger(config)
+
+        self.initialized = True
+
+    def get_parameters(self, config: Config) -> NDArrays:
+        """
+        Determines which parameters are sent back to the server for aggregation.
+            This uses a parameter exchanger to determine parameters sent.
+
+        For the ModelMergeClient, we assume that self.setup_client has already been called
+            as it does not support client polling so get_parameters is called from fit and
+            thus should be initialized by this point.
+
+        Args:
+            config (Config): The config is sent by the FL server to allow for customization in the function if desired.
+
+        Returns:
+            NDArrays: These are the parameters to be sent to the server. At minimum they represent the relevant model
+                parameters to be aggregated, but can contain more information.
+        """
+        assert self.model is not None
+        return self.parameter_exchanger.push_parameters(self.model, config=config)
+
+    def set_parameters(self, parameters: NDArrays, config: Config) -> None:
+        """
+        Sets the local model parameters transferred from the server using a parameter exchanger
+            to coordinate how parameters are set.
+
+        For the ModelMergeClient, we assume that initially parameters are being set to the parameters
+            in the nn.Module returned by the user defined get_model method. Thus, set_parameters is
+            only called once after model merging has occured and before federated evaluation.
+
+        Args:
+            parameters (NDArrays): Parameters have information about model state to be added to the relevant client
+                model but may contain more information than that.
+            config (Config): The config is sent by the FL server to allow for customization in the function if desired.
+            fitting_round (bool): Boolean that indicates whether the current federated learning round is a
+                fitting round or an evaluation round.
+                This is used to help determine which parameter exchange should be used for pulling parameters.
+                A full parameter exchanger is only used if the current federated learning round is the very
+                first fitting round.
+        """
+        assert self.initialized
+        self.parameter_exchanger.pull_parameters(parameters, self.model)
+
+    def fit(self, parameters: NDArrays, config: Config) -> Tuple[NDArrays, int, Dict[str, Scalar]]:
+        """
+        Initializes client, validates local client model on local test data and returns parameters,
+            test dataset length and test metrics. Importantly, parameters from Server, which is empty,
+            is not used to initialized the client model.
```

Comment:
Validate is called in other Client fit function, the only difference here is that we are calling on the test set instead of the validation. The rationale of doing the client side evaluation is to see how well the model performs on the local test set prior to model merging. We then distribute the merged model to the clients who again evaluate on the test set. Then we can see the difference in performance. I could make this optional, but then if no server_side_evaluation_fn is supplied (optional), there will be no metrics. Let me know if this makes sense to do and if I am missing something. I can definitely add some documentation that lays that out if you think its the right design decision

Reply:
Nope. That makes sense. Let's keep it like this and just add some discussion in the docstrings.